csi/src/error.rs

```diff
@@ -0,0 +1,290 @@
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by etcd_client::Error,
+    #[error("EtcdClientErr, the error is {:?}, context is {:?}", .source, .context)]
+    EtcdClientErr {
+        /// Error source
+        source: etcd_client::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by bincode::Error
+    #[error("BincodeErr, the error is {:?}, context is {:?}", .source, .context)]
+    BincodeErr {
+        /// Error source
+        source: bincode::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by nix::Error
+    #[error("NixErr, the error is {:?}, context is {:?}", .source, .context)]
+    NixErr {
+        /// Error source
+        source: nix::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Failed to mount
+    #[error("MountErr, fail to mount {:?} to {:?}, context is {:?}", .from, .target, .context)]
+    MountErr {
+        /// Source to mount
+        from: PathBuf,
+        /// Mount point
+        target: PathBuf,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Failed to umount
+    #[error("UmountErr, fail to umount {:?}, context is {:?}", .target, .context)]
+    UmountErr {
+        /// Mount point to umount
+        target: PathBuf,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by std::time::SystemTimeError
+    #[error("SystemTimeErr, the error is {:?}, context is {:?}", .source, .context)]
+    SystemTimeErr {
+        /// Error source
+        source: std::time::SystemTimeError,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by grpcio::Error
+    #[error("GrpcioErr, the error is {:?}, context is {:?}", .source, .context)]
+    GrpcioErr {
+        /// Error source
+        source: grpcio::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// API is not implemented
+    #[error("Not implemented, context is {:?}", .context)]
+    Unimplemented {
+        /// Context of the error
+        context: Vec<String>,
+    },
+}
+
+/// Add context to `DatenLordResult`
+pub trait Context<C, T> {
+    /// Add context to `DatenLordResult`
+    fn add_context(self, ctx: C) -> DatenLordResult<T>
+    where
+        C: Into<String>;
+}
+
+impl<C, T, E> Context<C, T> for Result<T, E>
+where
+    E: std::error::Error + Into<DatenLordError>,
+    C: Into<String>,
+{
+    fn add_context(self, ctx: C) -> DatenLordResult<T> {
+        self.map_err(|e| e.into().add_context(ctx))
+    }
+}
+
+impl DatenLordError {
+    /// Add context for `DatenLordError`
+    pub fn add_context<C>(mut self, ctx: C) -> Self
+    where
+        C: Into<String>,
+    {
+        macro_rules! append_context {
+            ($context: ident, [$($target:ident),*]) => {
```

Comment:
this `$context` is not used, right?

Reply:
`$context` matches the identifier `ctx` at line 214.

Reply:
Should change ctx to $context in line 208 to make more sense. Will modify.